### ISSUE TITLE
API: consistency in input validation

### ISSUE THREAD

**Issue by tupui** [scipy/scipy#18972]
### Is your feature request related to a problem? Please describe.

In #13761, I proposed a reflexion around input validation. The discussion was more focused on speed and in the end did not lead to anything.

This proposal is different and focuses on using a common infrastructure/tooling for input validation.

We currently don't have a convention when it comes to add input validation. Inputs are typically checked at the beginning of a function, with or without extra things around. And more recently in stats for instance, we have been trying to put all validations within a function with a pre/suffix `_iv`.

Regarding the validations we are also not consistent with checks that we are doing. We rarely check dtypes, sometimes check dimensions or we just forget to add some checks. When we have a specific check, we also see a wide range of implementations to do the same thing.

Lastly, our error messages are inconsistent. They are not formatted the same way, use the same formulation and don't have the same level of information.

### TL;DR
I propose to use a standardized data validation infrastructure. This would allow to share common validators, ensure input coverage, consistent error messages, lighter testing (possibly none if we trust the solutions), easier implementation, easier maintenance.

I investigated 2 solutions: a wrapper strategy; and the use of an external library (Pydantic).

### Wrapper
There is one solution close to us coming from Scikit-learn.

They are using a wrapper on top of the functions they want to validate the inputs from. Custom classes are used under the hood to do the validation.

### Pydantic
[Pydantic](https://docs.pydantic.dev) is one of the most widely used data validation library in the Python ecosystem. It's fast (Rust core), flexible, well maintained and funded and used as a dependency by many libraries like FastAPI.

Here the solution is not a wrapper, but uses classes similar to Data Classes and rely more on type annotations.

### Proof of Concept
I made 2 POC to show how this could work. Please be indulgent, these are POC meant to show that it works (tests are passing for both solutions) and how we could leverage some of the features of these options:

- https://github.com/tupui/scipy/pull/30
- https://github.com/tupui/scipy/pull/31

*On these POC, I did not measure a noticeable difference in terms of speed (between `main` and the 2 solutions)*

### Describe alternatives you've considered.

We could do nothing and try to document the way we want to do input validation. I argue that it's more work (learning the process, review, testing, maintenance down the road) for everyone than to use one of the proposed solution.

In #13761, I originally proposed to have 2 APIs. A fast one with validation and another one without validation. This idea can still be discussed and used in combination with any other idea around standardisation.

There is at list another widely used library to do input validation: `attrs`. I did not test this approach, mostly because I am not a user. Their `why not?` section is interesting (although most things they say or like to regarding Pydantic are not relevant/true anymore with Pydantic 2.0).

**Comment by mdhaber** [comment]
I prefer using the scikit-learn wrappers. This solution has less to learn, it seems to require less code to use than Pydantic (see @tupui's examples), the decorator approach stays completely out of the way of the function being validated, and the underlying library fits in a single file. 

It only attempts to do the simple input validation on each argument independently, since more complex Interdependent relationships are probably best handled with custom code anyway.

We could vendor it, or we could consider asking sklearn to split it into a separate repo and include it as a submodule. Either way, we don't have to worry about adding a dependency or users having older versions. I don't think maintenance burden would be unmanageable compared to what we do now, which is to maintain separate input validation code in almost every function. And since Pydantic doesn't seem to have special support for the needs of scientific Python, it's probably best to have an easy way to customizing it to our needs.

**Comment by dschmitz89** [comment]
I like the sklearn style input validation as well. Looks very simple to use and would make our validations much easier to maintain. Just thinking of all the different input validation ways of the different optimizations currently gives me chills..

**Comment by lorentzenchr** [comment]
I would distinguish 2 cases:
1. Input array validation
2. Input validation of all other parameters

At least the 2 POC PRs address the second point. For that one @jeremiedbb did a great job in https://github.com/scikit-learn/scikit-learn/pull/22722 and might answer questions about the scikit-learn solution better than me.

In scikit-learn, we also have the issue of nested (or repeated) input validation. This is solved by the `prefer_skip_nested_validation` argument of our solution.

Disclaimer: If have no experience with pydantic, but heard about it several times.

**Comment by lucascolley** [comment]
Copying a relevant comment after gh-19276 was closed:

---

Related to this is the possibility of optional parameters in `array_namespace` and `compliance_scipy`, or `as_xparray`, which could eventually eliminate the need for [`scipy._lib._util._asarray_validated`](https://github.com/scipy/scipy/blob/38410650cc0b0ffb058ec57b154a0bd688db8b8d/scipy/_lib/_util.py#L240-L293) (used in `linalg`). Here are its optional parameters:

```python
"""
...
check_finite : bool, optional
    Whether to check that the input matrices contain only finite numbers.
    Disabling may give a performance gain, but may result in problems
    (crashes, non-termination) if the inputs do contain infinities or NaNs.
    Default: True
sparse_ok : bool, optional
    True if scipy sparse matrices are allowed.
objects_ok : bool, optional
    True if arrays with dype('O') are allowed.
mask_ok : bool, optional
    True if masked arrays are allowed.
as_inexact : bool, optional
    True to convert the input array to a np.inexact dtype.
...
"""
```
